### PR TITLE
add ObservePipelineStateDuration

### PR DESCRIPTION
**Pull request by LenaHoinkis**
### Description

Suggestion for new metrics to identify pipeline run issues depending on their stages.
For example, you can then set an alert on the max value of the state. 
image
image

Difference of other metrics are:  
- I added a label for the pipeline run, so that we can also detect issues not only on average
- The time is always checked in the beginning of processing the run, not after state change, so that we can detect issues earlier

Tests are missing currently.

### Submitter checklist

- [ ] Change has been tested (on a back-end cluster)
- [ ] (If applicable) Jira backlog item ID added to the PR title and the [changelog.yaml] entry
- [ ] [changelog.yaml] entry with upgrade notes is prepared and appropriate for the audience affected by the change (users or developer, depending on the change)
- [ ] Semantic version diffed against [last release][releases] and updated accordingly. In this project the version has to be maintained here:
    - [/charts/steward/Chart.yaml](https://github.com/SAP/stewardci-core/blob/master/charts/steward/Chart.yaml) (`version` and `appVersion`)

In case dependencies have been updated:
- [ ] Links to external changelogs, since the last release of our component, added to the [changelog.yaml] entry (description).
- [ ] Changelogs read thoroughly, potential impact described, upgrade notes prepared (if necessary)
- [ ] Check if dependency updates affect our semantic version increment.

### Reviewer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There is at least 1 approval for the pull request and no outstanding requests for change
- [ ] All voter checks have passed
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] The Pull Request title is understandable and reflects the changes well
- [ ] The Pull Request description is understandable and well documented
- [ ] [changelog.yaml] entry for this Pull Request has been added
    - [ ] Changelog entry contains all required information
    - [ ] 'Upgrade notes' are documented in changelog.yaml (if required)

[changelog.yaml]: https://github.com/SAP/stewardci-core/changelog.yaml
[releases]: https://github.com/SAP/stewardci-core/releases

### CODE REVIEW

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -112,6 +120,18 @@ func (metrics *metrics) ObserveUpdateDurationByType(typ string, duration time.Du
 	metrics.Update.With(prometheus.Labels{"type": typ}).Observe(duration.Seconds())
 }
 
+func (metrics *metrics) ObservePipelineStateDuration(state *api.PipelineStatus, key string) error{
+	if state.StartedAt.IsZero() {
+		return fmt.Errorf("cannot observe StateItem if StartedAt is not set")
+	}
+	duration := time.Now().Sub(state.StartedAt.Time)
```
Comment (alxsap) [inline pkg/metrics/metrics.go:127]: does this (always) respect the time zone? Can `time.Now()` be in another time zone as `state.StartedAt.Time`? What would happen?

**Comment by alxsap** [comment]
Looks good from my perspective. Although I'm not much experienced with Prometheus.

`gofmt` fails in the voter pipeline.

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -64,6 +66,11 @@ func NewMetrics() Metrics {
 			Name: "steward_pipelineruns_total",
 			Help: "total number of pipelineruns",
 		}),
+		State: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "steward_state_duration",
+			Help: "duration of states by pipelines",
+		},
```
Comment (holgpar) [inline pkg/metrics/metrics.go:72]: I am not sure if I understand this correctly.
Do you intent to use a pipeline run identifer as label? Then you will generate lot's of metrics with very few data points. That is not a good design.
Metrics should be pre-aggregated by the application emitting them.

Above, a histogram was used to report about pipeline run duration. Why not follow the same approach to report about wait times?

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -64,6 +66,11 @@ func NewMetrics() Metrics {
 			Name: "steward_pipelineruns_total",
 			Help: "total number of pipelineruns",
 		}),
+		State: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "steward_state_duration",
+			Help: "duration of states by pipelines",
```
Comment (holgpar) [inline pkg/metrics/metrics.go:71]: Terminology: Pipeline vs Pipeline Run

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (holgpar) [reply pkg/metrics/metrics.go:127]: Timezones only come into play when convert a time object from it's internal representation into something human readable like hour, day, month...

**Comment by holgpar** [comment]
> I added a label for the pipeline run, so that we can also detect issues not only on average

As commented above, if you want to report about a single domain object you want to use either logging or an event store. Prometheus is not the right tool for this.

Applications have to pre-aggregate metrics. "Average" is only one possibility. If you use a histogram, you can alert as soon as there is one sample in a bad bucket.

**Comment by LenaHoinkis** [comment]
I will open an new PR with other approach

**Pull request closed without merging**